### PR TITLE
Add NFT-Based Voting & Dynamic Delegation Support to Votes.behavior.js

### PR DESCRIPTION
**Pull request by sohamsk13**
Dynamic vote weights based on NFT metadata

Proper delegation handling during transfers

Metadata changes affecting voting power

Batch operations for large collections

Edge cases around transfers and delegation

PR : 5209

### CODE REVIEW

**Comment by ernestognw** [comment]
Hi @sohamsk13, would you mind fixing the CI? If the tests are good we're happy to get them merged

**Comment by sohamsk13** [comment]
> Hi @sohamsk13, would you mind fixing the CI? If the tests are good we're happy to get them merged

Sure I will

File: test/governance/utils/Votes.behavior.js [hunk 1]
```diff
@@ -9,13 +9,34 @@ const { shouldBehaveLikeERC6372 } = require('./ERC6372.behavior');
...
   beforeEach(async function () {
-    [this.delegator, this.delegatee, this.alice, this.bob, this.other] = this.accounts;
+    [
+      this.delegator, 
+      this.delegatee, 
+      this.alice, 
+      this.bob, 
+      this.other,
+      this.charlie, 
+      this.dave
```
Comment (Amxx) [inline test/governance/utils/Votes.behavior.js:19]:
```suggestion
      this.charlie, 
      this.dave,
      this.other,
```

Comment (Amxx) [reply test/governance/utils/Votes.behavior.js:19]: I'd consider renaming to alice / bruce / chris / david, so that all names are 5 letters long

File: test/governance/utils/Votes.behavior.js [hunk 2]
```diff
@@ -9,13 +9,34 @@ const { shouldBehaveLikeERC6372 } = require('./ERC6372.behavior');
...
+    
+    // Mock metadata registry for NFT weight calculation
+    if (!fungible) {
+      this.metadata = {};
+      tokens.forEach(tokenId => {
+        this.metadata[tokenId] = { 
+          weight: (tokenId % 10n) + 1n, // Example: weight based on token ID
+          rarity: ['common', 'uncommon', 'rare', 'legendary'][Number(tokenId % 4n)] // Example metadata
+        };
+      });
```
Comment (Amxx) [inline test/governance/utils/Votes.behavior.js:31]: We can build the metadata registry directly
```suggestion
      this.metadata = Object.fromEntries(tokens.map(tokenId => [ 
        tokenId, 
        {
          weight: (tokenId % 10n) + 1n, // Example: weight based on token ID
          rarity: ['common', 'uncommon', 'rare', 'legendary'][Number(tokenId % 4n)] // Example metadata
        }
      ]));
```

File: test/governance/utils/Votes.behavior.js [hunk 3]
```diff
@@ -9,13 +9,34 @@ const { shouldBehaveLikeERC6372 } = require('./ERC6372.behavior');
...
+      
+      this.getNFTWeight = (tokenId) => this.metadata[tokenId]?.weight || 1n;
```
Comment (Amxx) [inline test/governance/utils/Votes.behavior.js:33]: we should not need that, `getWeight` (defined later) should be enough

File: test/governance/utils/Votes.behavior.js [hunk 4]
```diff
@@ -9,13 +9,34 @@ const { shouldBehaveLikeERC6372 } = require('./ERC6372.behavior');
...
+    }
   });
 
   shouldBehaveLikeERC6372(mode);
 
-  const getWeight = token => (fungible ? token : 1n);
+  const getWeight = token => (fungible ? token : (this.getNFTWeight ? this.getNFTWeight(token) : 1n));
```
Comment (Amxx) [inline test/governance/utils/Votes.behavior.js:39]:
```suggestion
  const getWeight = token => this?.metadata?.[token]?.weigth ?? 1n;
```
